homeassistant/components/life360/device_tracker.py

```diff
@@ -0,0 +1,617 @@
+    elif error_threshold and not warning_threshold:
+        # Make them the same which effectively prevents warnings.
+        config[CONF_WARNING_THRESHOLD] = error_threshold
+    else:
+        # Log all errors as errors.
+        config[CONF_ERROR_THRESHOLD] = 1
+        config[CONF_WARNING_THRESHOLD] = 1
+    return config
+
+
+_SLUG_LIST = vol.All(
+    cv.ensure_list, [cv.slugify],
+    vol.Length(min=1, msg='List cannot be empty'))
+
+_LOWER_STRING_LIST = vol.All(
+    cv.ensure_list, [vol.All(cv.string, vol.Lower)],
+    vol.Length(min=1, msg='List cannot be empty'))
+
+_EXCL_INCL_SLUG_LIST = vol.All(
+    vol.Schema({
+        vol.Exclusive(CONF_EXCLUDE, 'incl_excl'): _SLUG_LIST,
+        vol.Exclusive(CONF_INCLUDE, 'incl_excl'): _SLUG_LIST,
+    }),
+    cv.has_at_least_one_key(CONF_EXCLUDE, CONF_INCLUDE),
+    _excl_incl_list_to_filter_dict,
+)
+
+_EXCL_INCL_LOWER_STRING_LIST = vol.All(
+    vol.Schema({
+        vol.Exclusive(CONF_EXCLUDE, 'incl_excl'): _LOWER_STRING_LIST,
+        vol.Exclusive(CONF_INCLUDE, 'incl_excl'): _LOWER_STRING_LIST,
+    }),
+    cv.has_at_least_one_key(CONF_EXCLUDE, CONF_INCLUDE),
+    _excl_incl_list_to_filter_dict
+)
+
+_THRESHOLD = vol.All(vol.Coerce(int), vol.Range(min=1))
+
+LIFE360_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_USERNAME): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_ADD_ZONES): vol.Any(
+        AZ_ONCE, vol.All(cv.time_period, vol.Range(min=MIN_ZONE_INTERVAL))),
+    vol.Optional(CONF_CIRCLES): _EXCL_INCL_LOWER_STRING_LIST,
+    vol.Optional(CONF_DRIVING_SPEED): vol.Coerce(float),
+    vol.Optional(CONF_ERROR_THRESHOLD): _THRESHOLD,
+    vol.Optional(CONF_FILENAME, default=DEFAULT_FILENAME): cv.string,
+    vol.Optional(CONF_HOME_PLACE, default=DEFAULT_HOME_PLACE): cv.string,
+    vol.Optional(CONF_MAX_GPS_ACCURACY): vol.Coerce(float),
+    vol.Optional(CONF_MAX_UPDATE_WAIT): vol.All(
+        cv.time_period, cv.positive_timedelta),
+    vol.Optional(CONF_MEMBERS): _EXCL_INCL_SLUG_LIST,
+    vol.Optional(CONF_PLACES): _EXCL_INCL_LOWER_STRING_LIST,
+    vol.Optional(CONF_PREFIX, default=DEFAULT_PREFIX):
+        vol.All(vol.Any(None, cv.string), _prefix),
+    vol.Optional(CONF_SHOW_AS_STATE, default=[]): vol.All(
+        cv.ensure_list, [vol.In(SHOW_AS_STATE_OPTS)]),
+    vol.Optional(CONF_WARNING_THRESHOLD): _THRESHOLD,
+})
+
+PLATFORM_SCHEMA = vol.All(LIFE360_SCHEMA, _thresholds)
+
+
+def _include_name(filter_dict, name):
+    if not name:
+        return False
+    if not filter_dict:
+        return True
+    name = name.lower()
+    if filter_dict['include']:
+        return name in filter_dict['list']
+    return name not in filter_dict['list']
+
+
+def _exc_msg(exc):
+    return '{}: {}'.format(exc.__class__.__name__, str(exc))
+
+
+def _setup_zone_updating(hass, config, api):
+    from life360 import Life360Error
+
+    add_zones = config.get(CONF_ADD_ZONES)
+    if not add_zones:
+        return
+
+    Place = namedtuple(
```

Comment:
Move this outside of this function

Reply:
Can you explain why? It's only used inside this function.

Reply:
because you are defining a new class, and now are doing this each time this function is called.